public/components/notebooks/components/paragraph_components/paragraphs.tsx

```diff
@@ -105,25 +111,28 @@ export const Paragraphs = forwardRef((props: ParagraphProps, ref) => {
     http,
   } = props;
 
+  useEffect(() => console.log('Paragraphs', { para, index }), [index, para]);
+
   const [visOptions, setVisOptions] = useState<EuiComboBoxOptionOption[]>([]); // options for loading saved visualizations
   const [isPopoverOpen, setIsPopoverOpen] = useState(false);
   const [runParaError, setRunParaError] = useState(false);
   const [selectedVisOption, setSelectedVisOption] = useState<EuiComboBoxOptionOption[]>([]);
   const [visInput, setVisInput] = useState(undefined);
   const [visType, setVisType] = useState('');
+  const [textInput, setTextInput] = useState(props.para.inp);
 
   // output is available if it's not cleared and vis paragraph has a selected visualization
   const isOutputAvailable =
     (para.out.length > 0 && para.out[0] !== '') ||
     (para.isVizualisation && para.typeOut.length > 0 && visInput !== undefined);
 
   useImperativeHandle(ref, () => ({
-    runParagraph() {
-      return onRunPara();
+    runParagraph(evt) {
+      return onRunPara(evt);
     },
   }));
 
-  const fetchVisualizations = async () => {
+  const fetchVisualizations = useCallback(async () => {
```

Comment:
Many functions on Notebook and Paragraph consume and mutate multiple state values.  The common result of this over-complex state change is infinite-loops on state.
useCallback() buffers the change between the "changing thing" and the (temporarily) "constant thing".